### PR TITLE
WIP: show latitude/longitude

### PR DESCRIPTION
**Pull request by clausimu**
##### Reason for this pull request
I would like to add a latitude/longitude display to the player hud. This will be useful for any mission that requires finding or returning to a specific location on a planet. Currently, I use lat/lon for my search & rescue mission script where ships have crashed and the player is asked to go to the crash site and rescue personnel, deliver fuel, etc.

##### Feedback requested
I need some feedback on where the lat/lon info should be placed. Lat/Lon info is shown whenever the player is close enough to a planet to also show altitude. This pull request puts it in two places (just for testing) - centrally above the scanner, and on the right of the screen right above the panel (see screenshots below). The position above the scanner is problematic though because the relative speed info can run into it (see second screenshot).

lat_lon_3

lat_lon_5

### CODE REVIEW

**Comment by impaktor** [comment]
> The position above the scanner is problematic though because the relative speed info can run into it (see second screenshot).

Also, the different status bars could collide with it (hull/shield condition).

**Comment by clausimu** [comment]
> Also, the different status bars could collide with it (hull/shield condition).

Darn - I forgot about those! I don't want to put the coordinates randomly somewhere on the screen. But with the large scanner there is no room left at the bottom. I figure the left and right side of the screen should be left to different status bars, right?

**Comment by nozmajner** [comment]
Nice idea.
All four corners can be occupied by something as far as I know (or the top left may be free until there's a scanner equipment). How about moving the shield and hull bars up a bit to give room for the coordinates at the bottom right corner? They look right at that place for me, on your mockup.
With the compact scanner, it could look good above the altimeter I think. Not sure how it would behave with lower resolutions though.
Damn that OldUI flight display :)

**Comment by clausimu** [comment]
I will see what moving the status bars will do. Maybe I can fix the autopilot text that is running into the rotation damper buttons at the same time. Is anyone working on a better in-flight display? Or is what we currently have pretty much permanent?

**Comment by nozmajner** [comment]
Nobody currently. It was supposed to be ported to the newUI with the remaining oldUI screens.

**Comment by impaktor** [comment]
> How about moving the shield and hull bars up a bit to give room for the coordinates at the bottom right corner? They look right at that place for me, on your mockup.

:+1:

Also, I'd prefer if lat/long were formated in "deg,min,sec": 45° 13' 56''. (if we have "°" in our fonts?). This would be a function that goes into src/utils.cpp.

**Comment by fluffyfreak** [comment]
This is good, could be very handy.

**Comment by clausimu** [comment]
I have changed the coordinate display to degree/minute/second format and pushed the shield and hull bars up. The screenshot is trying to capture a situation where all bars are shown - in case you want me to align something differently.

lat_lon_6


I have also started to look into the flight UI a bit more closely but don't quite understand the philosophy of why which widget is setup in which file. It looks a little like it is stuck in the middle of a rewrite? Anyway, I decided to keep all things the same for now and use a new merge request for further changes.

**Comment by impaktor** [comment]
Time for nitpicking?

I think the green lat/log text needs slightly more space between text and blue buttons. Like the way "Set speed" has above/below to the blue buttons.

Also, I'm not sure what coding style we use here @fluffyfreak? In utils, it seems like we have utility functions in lowercase: `format_money`, `format_date`, and `format_distance`, but functions for converting types are in CamelCase: e.g. `FloatToStr` and `DoubleToStr`.

So question is, should it be `DecimalToDegMinSec` or `format_latlong` or `format_deg` or something similar?

@clausimu looks good otherwise. As I said, just some nitpicking on my part.

File: src/WorldView.cpp [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include "matrix4x4.h"
 #include "Quaternion.h"
 #include "LuaObject.h"
+//#include "utils.h"
```
Comment (impaktor) [inline src/WorldView.cpp]: Should this line be commented out?

**Comment by clausimu** [comment]
I increased the space a little bit and added more padding to the hull/shield bars on top to resemble the space on the left side between fuel and date. In general, I'm really not too happy about the arrangement of widgets. But that is something for another time :).

Nitpicking is very welcome! :)

lat_lon_7

**Comment by impaktor** [comment]
> I'm really not too happy about the arrangement of widgets.

I suspect no one is. They aught to be moved to a transparent Lua layer.

**Comment by walterar** [comment]
You can make a key activate Yes/No from LUA?

**Comment by impaktor** [comment]
@clausimu I guess you probably want to add that funtion to the LuaAPI. Check src/LuaFormat.cpp.

**Comment by impaktor** [comment]
So appart from this not being very useful atm, since no mission currently makes any use of it, is this done? Should we merge this with The Big Merge planned, or wait?

There sure are a lot of open pull requests at the moment.

File: src/WorldView.cpp [hunk 2]
```diff
@@ -823,9 +824,23 @@ void WorldView::RefreshButtonStateAndVisibility()
...
+                    m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_4, lon_str);
 				} else {
 					// XXX does this need to be repeated 3 times?
 					m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_BOTTOM_RIGHT, "");
+                    m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_1, "");
+                    m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_2, "");
+                    m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_3, "");
+                    m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_4, "");
 					if(m_curPlane != NONE) {
 						m_curPlane = NONE;
```
Comment (impaktor) [inline src/WorldView.cpp]: Indentation looks off here (L825-844). Maybe only visible when viewing with tab spaces 4 (append `?ts=4` to the github url to switch from tab width 8)

File: src/WorldView.cpp [hunk 3]
```diff
@@ -834,6 +849,10 @@ void WorldView::RefreshButtonStateAndVisibility()
 				}
 			} else {
 				m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_BOTTOM_RIGHT, "");
+                m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_1, "");
+                m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_2, "");
+                m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_3, "");
+                m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_OVER_PANEL_RIGHT_4, "");
 				if(m_curPlane != NONE) {
```
Comment (impaktor) [inline src/WorldView.cpp]: same here. Either the L836 / 851 is wrong or the ones that follow it.

**Comment by fluffyfreak** [comment]
I say we merge it with the BIG MERGE, formatting issues aside.

Missions and things using it can come along later.

**Comment by clausimu** [comment]
I will take a look to ensure correct formatting. What I haven't done yet is add the lat/long string formatting function to the LuaAPI (I'm using a local function for that in the search&rescue script).

**Comment by impaktor** [comment]
> I haven't done yet is add the lat/long string formatting function to the LuaAPI

That doesn't _have_ to be included in this PR, but if it's there, then it's in master and can be used for your Lua files without compiling the source. 

Just make sure you follow [our code style](http://pioneerwiki.com/wiki/Code_style) with tabs with 4 char width.

**Comment by clausimu** [comment]
Indentation issues fixed.

**Comment by walterar** [comment]
I think we should test it well. It is not finished.

**Comment by impaktor** [comment]
> I think we should test it well. It is not finished.

Is there anything specific to test? Do you feel like testing it / when?

I'm assigning this issue to me, since I suspect @fluffyfreak isn't too keen on rebasing, if I understood things correctly.

@clausimu I wonder why your commits aren't recognized by github as fully yours? Or why don't they have the same icon as your github account?

**Comment by fluffyfreak** [comment]
@walterar what's not finished about it? it's only displaying some formatted text on the HUD.

Thanks @impaktor yeah I'm going to struggle to get it done until later next week but will try to keep up!

**Comment by clausimu** [comment]
@impaktor I'm sorry - this has been my first experience with github. I'm sure there are still quite a few things I'm doing wrong. I just played around with my account settings and I think that fixed it.

@walterar Please let me know if you have any specific concerns I should test for/watch out for.

**Comment by impaktor** [comment]
> my first experience with github

Really? I thought you had some previous code merged a while back? Come by IRC if you want me to help you clean up your commit history.

> I think that fixed it.

Yes. Looks like expected now.

**Comment by clausimu** [comment]
> Really? I thought you had some previous code merged a while back? Come by IRC if you want me to help you clean up your commit history.

You are right - the forced time acceleration around ships. I meant whatever I have done with contributing to Pioneer. At some point I may have to take you up on the offer for help :)

**Comment by walterar** [comment]
I was working on this and found many "small" but important problems. For me it is easier to fix than explain. I'll think about how to do to convey to you the solutions. Today.

**Comment by impaktor** [comment]
> I'll think about how to do to convey to you the solutions

Give us the commit? Like so:

    git checkout -b show_lat_lon master
    git pull https://github.com/clausimu/pioneer.git show_lat_lon

Make your changes, then `git add <file>` and `git commit -m "Fixes placement of lat long on HUD"` then push to your github, 

    git push origin show_lat_lon

(assuming you're github is called `origin` on your computer)
and link to the commit here, so I can cherry-pick it and include it in this pull request.

**Comment by walterar** [comment]
@impaktor I have to learn to handle this directly with a branch on Github, as you suggest, but not today.  Please, do it for me.

It is this zip file, https://dl.dropboxusercontent.com/u/15391580/Pioneer/geopos.zip find the source files with the changes applied, more executables for Windows 32 and Linux 64 proposed, together with the executables of this PR (3499)

Executable files are renamed, to avoid overwriting the original.

The language files, data/lang/core/en.json, is also in a mod file

I think this, everyone can test, one and other.

If something is not understood, let me know.

**Comment by impaktor** [comment]
So you moved the lat lon to the bottom? Is this what we want? @nozmajner ?

screenshot-20151030-185151

screenshot-20151030-185208

I'll leave it to @clausimu to look at the diff from your dropbox.

**Comment by walterar** [comment]
All this work, I did over Scout+
Here some screenshot.

geo1
geo2

You may want to improve and change things for Pioneer vanilla

**Comment by impaktor** [comment]
So the problem you said, is that @clausimu's changes look bad in scout+?

@clausimu if you have time and want to play around more with pioneer: 
- I had this idea that the x,y position of all our buttons on the HUD could be set in a data/ui/*.ini-file. That way nozmajner could play around with button placement which he wanted to do, and walterar can easily change scout+, or perhaps this idea of mine has problems? Either way, this is just something I thought of. I don't know whay @fluffyfreak thinks about it?

- Also, the /data/icons/multifunc_equip_on.png and /data/icons/multifunc_scanner_on.png could be removed and be one single button which flipps between the two "off" version of them. Saves us space, and would be consistent with how other buttons behave.

Anyway, I don't have time to play with that, so I thought someone else might want to have a go at it, shouldn't be too difficult I think.

**Comment by nozmajner** [comment]
I'm not really liking them under the scanner. One thing is that they overlap the compact scanner.
Also, it overlaps the missile selector a bit too.

Also I think that they look much better and cleaner on the right. On the bottom, they look like if they are randomly placed. At least with this to the slightly right placement, it looks hackish.

**Comment by walterar** [comment]
I have yet to reduce some the compact scanner, but I think will removed.

**Comment by nozmajner** [comment]
So the coordinates doesn't hide when you aren't close to a planet/hyperjump it seem.

**Comment by impaktor** [comment]
> So the coordinates doesn't hide when you aren't close to a planet/hyperjump it seem. 

They "hide" at the same time as the two directional indicators.

**Comment by nozmajner** [comment]
They haven't after a hyperjump.

**Comment by walterar** [comment]
wrong-site
wrong-site2
wrong-site3
wrong-site4

**Comment by walterar** [comment]
"Also, the /data/icons/multifunc_equip_on.png and /data/icons/multifunc_scanner_on.png could be removed and be one single button which flipps between the two "off" version of them. Saves us space, and would be consistent with how other buttons behave."

In a combat with 4 or 5 pirates, launching missiles with the mouse is as useful as an ashtray on a motorbike. ;)

**Comment by fluffyfreak** [comment]
> I had this idea that the x,y position of all our buttons on the HUD could be set in a data/ui/*.ini-file. 
> That way nozmajner could play around with button placement which he wanted to do, and walterar 
> can easily change scout+, or perhaps this idea of mine has problems? Either way, this is just 
> something I thought of. I don't know whay @fluffyfreak thinks about it?

I'd prefer we converted it to new-UI than add another way of editing the existing hud...

My personal preference right now is that they Lat/Lon should go to the right or left of the scanner, above the Pressure or Altitude readings. It does intrude on the wide scanner but I don't care as it's mostly to the edge of it.

Also like the Altitude and Pressure it should disappear / appear when they do.

However, I want these things and so if @clausimu would prefer then we can merge this first and I will make the changes?
Options therefore are:
- right of screen above icons,
- left of scanner above Pressure,
- right of scanner above Altitude,
- bottom beneath scanner.

If you want a say then no long replies just one of those. We can change the position in the future, or someone can open another PR/issue about the placement for us to decide.

**Comment by walterar** [comment]
If you see my code, they will realize that the aesthetic is not the only thing that changed.

Regarding aesthetics: I tried all positions, for many hours of play, and this, is the best I saw. I made some final adjustments to the position of the icons of missiles and dimensions of the compact radar screen. That is how come out in Scout+

missilepos
radarpos

In future, I would like to have new equipment called GPS. :)

Are free to do whatever they want. And if you want, do not worry about the credit; It is a collaboration of Scout+

**Comment by impaktor** [comment]
> I'd prefer we converted it to new-UI than add another way of editing the existing hud...

Sure. It's just that I won't touch that project with a 50 foot pole (nor will anyone else), since I believe it's vastly beyond my capabilities. (And what I can motivate myself to learn).

> Also like the Altitude and Pressure it should disappear / appear when they do.

Yes, and if it doesn't after hyperspace then that's a bug that aught to be fixed before merge.

> My personal preference right now is that they Lat/Lon should go to the right or left of the scanner

Poor @clausimu getting 2 persons saying one thing, and 2 another :smile:. I share nozmajners opinion, so, I guess I'm quite biased in suggesting "let nozmajner have the final call". I don't feel strongly about where it should go. Whichever we decide, we could ask for player feedback.

> GPS

Actually, this could be a 0t equipment, easy to add.

> If you see my code, they will realize that the aesthetic is not the only thing that changed.

Position and mouse over text is the only thing I noticed, but I had trouble reading the diff.

**Comment by laarmen** [comment]
PPS: Planetary positioning system. And since it's supposed to work
everywhere, it ought to be at least 100kg ;-)

**Comment by laarmen** [comment]
Also, I have no opinion on the matter, except that noz is probably right
:-)

**Comment by fluffyfreak** [comment]
Ok, we'll leave it as @clausimu has done it now.

I've branched from this PR but can't seem to create a PR into @clausimu branch with my fixes.
My branch is [fluffyfreak:show_lat_lon](https://github.com/fluffyfreak/pioneer/tree/show_lat_lon) with two additional commits that fix the post-hyperspace issue, and some minor indentation.

**Comment by fluffyfreak** [comment]
I think you can manually bring them across using something like the following:
> git fetch https://github.com/fluffyfreak/pioneer.git
> git cherry-pick 74112b71ec308ba7ae272378f5a87e0d4fa8aa14
> git cherry-pick 288b9b7e8b8e057c901a291dbcb6ea71999ccd08

But it'd be good if @impaktor / @laarmen could double check those commends for me!

**Comment by walterar** [comment]
"but I had trouble reading the diff."

Long ago I use Gnome Commander, combined with Meld. It's my main tool. I like doing all mergers "artisanal" way. It allows me to have total control of what I am doing and also get familiar with the code.

**Comment by walterar** [comment]
"Ok, we'll leave it as @clausimu has done it now."

Ok, then I withdraw my offer.

**Comment by impaktor** [comment]
I had a thought regarding having the GPS/PPS equipment. I think this would go very well with being `purchasable=false`, and instead, upon doing the first search-and-rescue mission, being asked to buy one from the contractor.
- GPS as equipment is good because more equipment = more upgrades = more fun/addiction
- But putting an obstacle in front of getting one is also good, it makes the player strive to get/find a rescue mission, so he can "complete" his/her ship, and get a very exclusive piece of equip. that can't "just be bought at the station"
- So in summary: think of Frontier where you had to do a reconnaissance mission to get a camera. Basically like that.
